File: DevExpress VCL/ExpressQuantumTreeList 4/Demos/Delphi/StylesMultiDemo/StylesMultiDemoMain.pas
unit StylesMultiDemoMain;

interface

uses
  Windows, Messages, Forms, SysUtils, cxStyles, cxCustomData, cxGraphics, cxFilter,
  cxEdit, DB, cxDBData, Dialogs,Classes, ActnList, ImgList, Controls, Menus,
  StdCtrls, ExtCtrls, ShellAPI, cxButtons, cxData, cxClasses, cxControls,
  ComCtrls, cxLookAndFeelPainters, cxContainer, cxRadioGroup, cxListBox, cxTL,
  DemoBasicMain, cxLookAndFeels, cxMaskEdit, cxMemo, cxCurrencyEdit,
  cxCheckBox, cxDBLookupComboBox, cxInplaceContainer, cxDBTL, cxTLData,
  cxTextEdit, cxCalc;

type
  TcxStyleRepositoryType = (shtNone, shtPredefined, shtUserDefined);

  TStylesMultiDemoMainForm = class(TDemoBasicMainForm)
    miGridLookFeel: TMenuItem;
    miKind: TMenuItem;
    miFlat: TMenuItem;
    miStandard: TMenuItem;
    miUltraFlat: TMenuItem;
    miNativeStyle: TMenuItem;
    miSeparator2: TMenuItem;
    pnlLeft: TPanel;
    Splitter: TSplitter;
    actSaveToFile: TAction;
    OpenDialog: TOpenDialog;
    SaveDialog: TSaveDialog;
    actLoadFromFile: TAction;
    gbUserDefined: TGroupBox;
    gbPredefined: TGroupBox;
    btnLoad: TcxButton;
    btnSave: TcxButton;
    btnEdit: TcxButton;
    actEditStyleSheet: TAction;
    pnlCurrentStyleSheet: TPanel;
    cxDBTreeList: TcxDBTreeList;
    cxDBTreeListID: TcxDBTreeListColumn;
    cxDBTreeListPARENTID: TcxDBTreeListColumn;
    cxDBTreeListNAME: TcxDBTreeListColumn;
    cxDBTreeListPHONE: TcxDBTreeListColumn;
    cxDBTreeListFAX: TcxDBTreeListColumn;
    cxDBTreeListBUDGET: TcxDBTreeListColumn;
    cxDBTreeListVACANCY: TcxDBTreeListColumn;
    cxDBTreeListManager: TcxDBTreeListColumn;
    cxDBTreeListManagerPhone: TcxDBTreeListColumn;
    cxDBTreeListManagerEmail: TcxDBTreeListColumn;
    cxDBTreeListManagerAdress: TcxDBTreeListColumn;
    tlStyleSheets: TcxTreeList;
    clnRadio: TcxTreeListColumn;
    clnGroupName: TcxTreeListColumn;
    Office111: TMenuItem;
    procedure LookAndFeelChange(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure actSaveToFileExecute(Sender: TObject);
    procedure actLoadFromFileExecute(Sender: TObject);
    procedure actEditStyleSheetExecute(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure FormActivate(Sender: TObject);
    procedure tlStyleSheetsIsGroupNode(Sender: TObject;
      ANode: TcxTreeListNode; var IsGroup: Boolean);
    procedure tlStyleSheetsStylesGetContentStyle(Sender, AItem: TObject;
      ANode: TcxTreeListNode; var AStyle: TcxStyle);
    procedure tlStyleSheetsStylesGetNodeIndentStyle(Sender: TObject;
      ANode: TcxTreeListNode; ALevel: Integer; var AStyle: TcxStyle);
    procedure tlStyleSheetsSelectionChanged(Sender: TObject);
    procedure cxDBTreeListInitInsertingRecord(Sender: TObject;
      AFocusedNode: TcxTreeListDataNode; var AHandled: Boolean);
    procedure cxDBTreeListDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
  private
    function GetCurrentStyleSheet: TcxTreeListStyleSheet;
    procedure CreateStyleSheetsList(AStyleRepositoryType: TcxStyleRepositoryType);
    procedure UpdateGridStyleSheets(const AStyleSheet: TcxTreeListStyleSheet);
    procedure ChangeVisibility(AType: TcxStyleRepositoryType);
    procedure ClearUserDefinedStyleSheets;
    procedure LoadUserDefinedStyleSheets(AFileName: TFileName);
    procedure SaveUserDefinedStyleSheets(AFileName: TFileName);
    procedure SelectFistChild(AStyleRepositoryType: TcxStyleRepositoryType);
  end;

var
  StylesMultiDemoMainForm: TStylesMultiDemoMainForm;

implementation

uses
  StylesMultiDemoData, cxStyleSheetEditor, cxTLStyleSheetPreview;

{$R *.dfm}

procedure TStylesMultiDemoMainForm.LookAndFeelChange(Sender: TObject);
begin
  if TMenuItem(Sender).Tag > 3 then
  begin
    cxLookAndFeelController.NativeStyle :=
      not cxLookAndFeelController.NativeStyle;
    TMenuItem(Sender).Checked := cxLookAndFeelController.NativeStyle;
  end
  else
  begin
    TMenuItem(Sender).Checked := True;
    cxLookAndFeelController.Kind := TcxLookAndFeelKind(TMenuItem(Sender).Tag);
    cxLookAndFeelController.NativeStyle := False;
    miNativeStyle.Checked := False;
  end;
end;

procedure TStylesMultiDemoMainForm.FormCreate(Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  with StylesMultiDemoDataDM do
  begin
    CreateStyleSheetsList(shtUserDefined);
    CreateStyleSheetsList(shtPredefined);
  end;
  SelectFistChild(shtPredefined);
  cxDBTreeList.FullExpand;

//}
end;

procedure TStylesMultiDemoMainForm.UpdateGridStyleSheets(const AStyleSheet: TcxTreeListStyleSheet);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  if GetCurrentStyleSheet = AStyleSheet then Exit;
  cxDBTreeList.Styles.StyleSheet := AStyleSheet;
  if AStyleSheet <> nil then
    pnlCurrentStyleSheet.Caption := AStyleSheet.Caption
  else
    pnlCurrentStyleSheet.Caption := 'None';
  cxDBTreeList.Update;
//}
end;

procedure TStylesMultiDemoMainForm.actSaveToFileExecute(Sender: TObject);
begin
  with SaveDialog do
    if Execute then
      SaveUserDefinedStyleSheets(FileName);
end;

procedure TStylesMultiDemoMainForm.actLoadFromFileExecute(Sender: TObject);
begin
  with OpenDialog do
    if Execute then
    begin
      LoadUserDefinedStyleSheets(FileName);
      SelectFistChild(shtUserDefined);
    end;
end;

procedure TStylesMultiDemoMainForm.ChangeVisibility(AType: TcxStyleRepositoryType);
begin
  btnEdit.Enabled := AType = shtUserDefined;
  btnSave.Enabled := AType = shtUserDefined;
end;

procedure TStylesMultiDemoMainForm.actEditStyleSheetExecute(
  Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  ShowcxStyleSheetEditor(GetCurrentStyleSheet, nil);

//}
end;

procedure TStylesMultiDemoMainForm.FormShow(Sender: TObject);
begin
{ remove/add the closing brace on this line to disable/enable the following code

  ShowMessage('WARNING: tutorial not completed. First, please apply the steps '+
              'shown in the StylesMultiDemo.doc file');

//}
end;

function TStylesMultiDemoMainForm.GetCurrentStyleSheet: TcxTreeListStyleSheet;
begin
  Result := TcxTreeListStyleSheet(cxDBTreeList.Styles.StyleSheet);
end;

procedure TStylesMultiDemoMainForm.LoadUserDefinedStyleSheets(AFileName: TFileName);
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  UpdateGridStyleSheets(nil);
  ClearUserDefinedStyleSheets;

  LoadStyleSheetsFromIniFile(AFileName, StylesMultiDemoDataDM.strepUserDefined,
    TcxTreeListStyleSheet);

  CreateStyleSheetsList(shtUserDefined);
//}
end;

procedure TStylesMultiDemoMainForm.SaveUserDefinedStyleSheets(AFileName: TFileName);
var
  AList: TList;
begin
{ remove/add the closing brace on this line to disable/enable the following code}

  AList := TList.Create;
  try
    PopulateStyleSheetsList(AList);
    SaveStyleSheetsToIniFile(AFileName, AList);
   finally
      AList.Free;
    end;

//}
end;

procedure TStylesMultiDemoMainForm.ClearUserDefinedStyleSheets;
begin
  with StylesMultiDemoDataDM.strepUserDefined do
  begin
    Clear;
    ClearStyleSheets;
  end;
end;

procedure TStylesMultiDemoMainForm.FormActivate(Sender: TObject);
begin
  OpenDialog.InitialDir := ExtractFileDir(Application.ExeName);
  SaveDialog.InitialDir := OpenDialog.InitialDir;
end;

procedure TStylesMultiDemoMainForm.tlStyleSheetsIsGroupNode(
  Sender: TObject; ANode: TcxTreeListNode; var IsGroup: Boolean);
begin
  if (ANode.Level = 0) and (ANode.Index <> Integer(shtNone)) then
   IsGroup := True;
end;

procedure TStylesMultiDemoMainForm.tlStyleSheetsStylesGetContentStyle(
  Sender, AItem: TObject; ANode: TcxTreeListNode; var AStyle: TcxStyle);
begin
  with StylesMultiDemoDataDM do
    if ANode.Level = 0 then
      AStyle := styGroupNode
    else
      AStyle := styOptionalNode;
end;

procedure TStylesMultiDemoMainForm.tlStyleSheetsStylesGetNodeIndentStyle(
  Sender: TObject; ANode: TcxTreeListNode; ALevel: Integer;
  var AStyle: TcxStyle);
begin
  with StylesMultiDemoDataDM do
    AStyle := styGroupNode;
end;

procedure TStylesMultiDemoMainForm.tlStyleSheetsSelectionChanged(
  Sender: TObject);
  procedure SetCheckedNode(ANode: TcxTreeListNode);
  var
    I: Integer;
  begin
    if ANode <> nil then
      ANode.Values[0] := 1;
    with tlStyleSheets do
      for I := 0 to Nodes.Count - 1 do
        if Nodes[I] <> ANode then
          Nodes[I].Values[clnRadio.ItemIndex] := False;
  end;

var
  ANode: TcxTreeListNode;
begin
  if tlStyleSheets.SelectionCount > 0 then
  begin
    ANode := tlStyleSheets.Selections[0];
    if not ANode.IsGroupNode then
    begin
      SetCheckedNode(ANode);
      UpdateGridStyleSheets(TcxTreeListStyleSheet(ANode.Data));
      ANode.Values[clnRadio.ItemIndex] := True;
      ChangeVisibility(TcxStyleRepositoryType(ANode.Parent.Index));
    end;
    tlStyleSheets.Update;
  end;
end;

procedure TStylesMultiDemoMainForm.CreateStyleSheetsList(
  AStyleRepositoryType: TcxStyleRepositoryType);
var
  I: Integer;
  ANode: TcxTreeListNode;
  AStyleRepository: TcxStyleRepository;
begin
  if AStyleRepositoryType = shtNone then
    Exit;
  with StylesMultiDemoDataDM do
    if AStyleRepositoryType = shtUserDefined then
      AStyleRepository := strepUserDefined
    else
      AStyleRepository := strepPredefined;
  with AStyleRepository do
  begin
    ANode := tlStyleSheets.Nodes.Root[Integer(AStyleRepositoryType)];
    ANode.DeleteChildren;
    for I := 0 to StyleSheetCount - 1 do
      with ANode.AddChild do
      begin
        Values[clnGroupName.ItemIndex] := StyleSheets[I].Caption;
        Data := StyleSheets[I];
      end;
  end;
end;
procedure TStylesMultiDemoMainForm.SelectFistChild(
  AStyleRepositoryType: TcxStyleRepositoryType);
var
  ANode: TcxTreeListNode;
begin
  ANode := tlStyleSheets.Nodes.Root[Integer(AStyleRepositoryType)];
  if ANode.Count > 0 then
  begin
    ANode[0].Focused := True;
    ANode[0].MakeVisible;
  end;
end;

procedure TStylesMultiDemoMainForm.cxDBTreeListInitInsertingRecord(
  Sender: TObject; AFocusedNode: TcxTreeListDataNode;
  var AHandled: Boolean);
begin
  if AFocusedNode <> nil then
    StylesMultiDemoDataDM.SetParentValue(AFocusedNode.ParentValue);
end;

procedure TStylesMultiDemoMainForm.cxDBTreeListDragOver(Sender,
  Source: TObject; X, Y: Integer; State: TDragState; var Accept: Boolean);
begin
  inherited;
  //
end;

end.

File: DevExpress VCL/ExpressQuantumTreeList 4/Demos/CBuilder/StylesMultiDemo/StylesMultiDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "StylesMultiDemoMain.h"
#include "StylesMultiDemoData.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxButtons"
#pragma link "cxCalc"
#pragma link "cxCheckBox"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDBTL"
#pragma link "cxInplaceContainer"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxMaskEdit"
#pragma link "cxMemo"
#pragma link "cxRadioGroup"
#pragma link "cxTextEdit"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma link "cxStyleSheetEditor"
#pragma link "cxTLStyleSheetPreview"
#pragma resource "*.dfm"
TStylesMultiDemoMainForm *StylesMultiDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TStylesMultiDemoMainForm::TStylesMultiDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::FormShow(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code *

  ShowMessage("WARNING: tutorial not completed. First, please apply the steps "
    "shown in the doc file");

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::LookAndFeelChange(TObject *Sender)
{
  if (((TMenuItem*)Sender)->Tag > 3){
    cxLookAndFeelController->NativeStyle =
      !cxLookAndFeelController->NativeStyle;
    ((TMenuItem*)Sender)->Checked = cxLookAndFeelController->NativeStyle;
  }
  else{
    ((TMenuItem*)Sender)->Checked = true;
    cxLookAndFeelController->Kind = (TcxLookAndFeelKind)((TMenuItem*)Sender)->Tag;
    cxLookAndFeelController->NativeStyle = false;
    miNativeStyle->Checked = false;
  }
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::FormCreate(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  CreateStyleSheetsList(shtUserDefined);
  CreateStyleSheetsList(shtPredefined);
  SelectFistChild(shtPredefined);
  cxDBTreeList->FullExpand();

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::actSaveToFileExecute(TObject *Sender)
{
  if (SaveDialog->Execute())
    SaveUserDefinedStyleSheets(SaveDialog->FileName);
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::actLoadFromFileExecute(TObject *Sender)
{
  if (OpenDialog->Execute()){
    LoadUserDefinedStyleSheets(OpenDialog->FileName);
    SelectFistChild(shtUserDefined);
  }
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::actEditStyleSheetExecute(TObject *Sender)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  ShowcxStyleSheetEditor(GetCurrentStyleSheet(), NULL);

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::FormActivate(TObject *Sender)
{
  OpenDialog->InitialDir = ExtractFileDir(Application->ExeName);
  SaveDialog->InitialDir = OpenDialog->InitialDir;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::tlStyleSheetsIsGroupNode(
  TObject *Sender, TcxTreeListNode *ANode, bool &IsGroup)
{
  if (ANode->Level == 0 && ANode->Index != (int)shtNone)
   IsGroup = true;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::tlStyleSheetsStylesGetContentStyle(
  TObject *Sender, TObject *AItem, TcxTreeListNode *ANode, TcxStyle *&AStyle)
{
  if (ANode->Level == 0)
    AStyle = StylesMultiDemoDataDM->styGroupNode;
  else
    AStyle = StylesMultiDemoDataDM->styOptionalNode;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::tlStyleSheetsStylesGetNodeIndentStyle(
  TObject *Sender, TcxTreeListNode *ANode, int ALevel, TcxStyle *&AStyle)
{
  AStyle = StylesMultiDemoDataDM->styGroupNode;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::tlStyleSheetsSelectionChanged(
  TObject *Sender)
{
  if (tlStyleSheets->SelectionCount > 0){
    TcxTreeListNode *ANode = tlStyleSheets->Selections[0];
    if (!ANode->IsGroupNode){
      SetCheckedNode(ANode);
      UpdateGridStyleSheets((TcxTreeListStyleSheet*)ANode->Data);
      ANode->Values[clnRadio->ItemIndex] = true;
      ChangeVisibility((TcxStyleRepositoryType)ANode->Parent->Index);
    }
    tlStyleSheets->Update();
  }
}

void __fastcall TStylesMultiDemoMainForm::SetCheckedNode(TcxTreeListNode *ANode)
{
    if (ANode != NULL)
      ANode->Values[0] = 1;
    for (int I = 0; I < tlStyleSheets->Nodes->Count; I++)
      if (tlStyleSheets->Nodes->Items[I] != ANode)
        tlStyleSheets->Nodes->Items[I]->Values[clnRadio->ItemIndex] = false;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::cxDBTreeListInitInsertingRecord(
  TObject *Sender, TcxTreeListDataNode *AFocusedNode, bool &AHandled)
{
  if (AFocusedNode != NULL)
    StylesMultiDemoDataDM->SetParentValue(AFocusedNode->ParentValue);
}
//---------------------------------------------------------------------------

TcxTreeListStyleSheet* __fastcall TStylesMultiDemoMainForm::GetCurrentStyleSheet()
{
  return (TcxTreeListStyleSheet*)cxDBTreeList->Styles->StyleSheet;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::CreateStyleSheetsList(
  TcxStyleRepositoryType AStyleRepositoryType)
{
  if (AStyleRepositoryType == shtNone)
    return;
  TcxStyleRepository *AStyleRepository;
  if (AStyleRepositoryType == shtUserDefined)
    AStyleRepository = StylesMultiDemoDataDM->strepUserDefined;
  else
    AStyleRepository = StylesMultiDemoDataDM->strepPredefined;
  TcxTreeListNode *ANode = tlStyleSheets->Nodes->Root->Items[(int)AStyleRepositoryType];
  ANode->DeleteChildren();
  TcxTreeListNode *AStyleNode;
  for (int I = 0; I < AStyleRepository->StyleSheetCount; I++){
    AStyleNode = ANode->AddChild();
    AStyleNode->Values[clnGroupName->ItemIndex] =
      AStyleRepository->StyleSheets[I]->Caption;
    AStyleNode->Data = AStyleRepository->StyleSheets[I];
  }
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::UpdateGridStyleSheets(
  TcxTreeListStyleSheet *AStyleSheet)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  if (GetCurrentStyleSheet() == AStyleSheet) return;
  cxDBTreeList->Styles->StyleSheet = AStyleSheet;
  if (AStyleSheet != NULL)
    pnlCurrentStyleSheet->Caption = AStyleSheet->Caption;
  else
    pnlCurrentStyleSheet->Caption = "None";
  cxDBTreeList->Update();

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::ChangeVisibility(
  TcxStyleRepositoryType AType)
{
  btnEdit->Enabled = AType == shtUserDefined;
  btnSave->Enabled = AType == shtUserDefined;
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::ClearUserDefinedStyleSheets()
{
  StylesMultiDemoDataDM->strepUserDefined->Clear();
  StylesMultiDemoDataDM->strepUserDefined->ClearStyleSheets();
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::LoadUserDefinedStyleSheets(
  TFileName AFileName)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  UpdateGridStyleSheets(NULL);
  ClearUserDefinedStyleSheets();

  LoadStyleSheetsFromIniFile(AFileName, StylesMultiDemoDataDM->strepUserDefined,
    __classid(TcxTreeListStyleSheet), NULL, NULL, NULL, NULL);

  CreateStyleSheetsList(shtUserDefined);

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::SaveUserDefinedStyleSheets(
  TFileName AFileName)
{
/* remove/add the closing slash on this line to disable/enable the following code */

  TList *AList = new TList();
  __try{
    PopulateStyleSheetsList(AList);
    SaveStyleSheetsToIniFile(AFileName, AList);
  }
  __finally{
    delete AList;
  }

//*/
}
//---------------------------------------------------------------------------

void __fastcall TStylesMultiDemoMainForm::SelectFistChild(
  TcxStyleRepositoryType AStyleRepositoryType)
{
  TcxTreeListNode *ANode =
    tlStyleSheets->Nodes->Root->Items[(int)AStyleRepositoryType];
  if (ANode->Count > 0){
    ANode->Items[0]->Focused = true;
    ANode->Items[0]->MakeVisible();
  }
}
//---------------------------------------------------------------------------


void __fastcall TStylesMultiDemoMainForm::cxDBTreeListDragOver(
      TObject *Sender, TObject *Source, int X, int Y, TDragState State,
      bool &Accept)
{
//        
}
//---------------------------------------------------------------------------

